local/ownDevice.go

```
@@ -0,0 +1,496 @@
+package local
+
+import (
+	"context"
+
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+
+	"sync"
+	//"encoding/base64"
+
+	gocoap "github.com/go-ocf/go-coap"
+	"github.com/go-ocf/sdk/local/device"
+	"github.com/go-ocf/sdk/local/resource"
+	"github.com/go-ocf/sdk/schema"
+)
+
+type deviceOwnershipClient struct {
+	*coapClient
+	ownership schema.Doxm
+}
+
+func (c *deviceOwnershipClient) GetOwnership() schema.Doxm {
+	return c.ownership
+}
+
+func (c *deviceOwnershipClient) GetTcpSecureAddress(ctx context.Context, deviceID string) (string, error) {
+	deviceLink, err := c.GetDeviceLinks(ctx, deviceID)
+	if err != nil {
+		return "", err
+	}
+	var resourceLink schema.ResourceLink
+	for _, link := range deviceLink.Links {
+		if link.HasType("oic.wk.d") {
+			resourceLink = link
+			break
+		}
+	}
+	tcpTLSAddr, err := resourceLink.GetTCPSecureAddr()
+	if err != nil {
+		return "", err
+	}
+	return tcpTLSAddr.String(), nil
+}
+
+type deviceOwnershipHandler struct {
+	deviceID string
+	cancel   context.CancelFunc
+
+	client *deviceOwnershipClient
+	lock   sync.Mutex
+	err    error
+}
+
+func newDeviceOwnershipHandler(deviceID string, cancel context.CancelFunc) *deviceOwnershipHandler {
+	return &deviceOwnershipHandler{deviceID: deviceID, cancel: cancel}
+}
+
+func (h *deviceOwnershipHandler) Handle(ctx context.Context, clientConn *gocoap.ClientConn, ownership schema.Doxm) {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+	if ownership.DeviceId == h.deviceID && h.client == nil {
+		h.client = &deviceOwnershipClient{coapClient: NewCoapClient(clientConn, schema.UDPScheme), ownership: ownership}
+		h.cancel()
+	}
+}
+
+func (h *deviceOwnershipHandler) Error(err error) {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+	h.err = err
```

Can you fail fast and prevent overwriting the initial error?